### PR TITLE
feat(api): interface and helpers for importing datasources

### PR DESCRIPTION
**Pull request by sundersc**
Adds interfaces and helper classes required to construct internal data representation for importing existing data sources.

### CODE REVIEW

**Review by phani-srikar — APPROVED** [review]

File: packages/amplify-graphql-schema-generator/src/schema-representation/engine.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export type DBEngineType = 'MySQL' | 'DynamoDB';
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/engine.ts:1]: Is MySQL the right granulatity here, or do we want 'Relational' vs. 'NoSQL'? or something like that?

File: packages/amplify-graphql-schema-generator/src/schema-representation/schema.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    return this.models;
+  }
+
+  public getEngine(): Engine {
+    return this.engine;
+  }
+
+  public addModel(model: Model): void {
+    if (this.hasModel(model.getName())) {
+      throw new Error("Model already exists");
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/schema.ts:19]: Model Name in the error?

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+export type FieldType = DefaultType | CustomType | ListType | NonNullType;
+
+export interface DefaultType {
+  readonly kind: 'Scalar';
+  readonly name: 'String' | 'ID' | 'Int' | 'Float' | 'AWSJSON' | 'AWSDate' | 'AWSTime' | 'AWSDateTime' | 'AWSTimestamp';
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:5]: Boolean, AWSEmail, AWSPhone, AWSURL, AWSIPAddress are missing.

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+export interface NonNullType {
+  readonly kind: 'NonNull';
+  readonly type: DefaultType | CustomType | ListType;
+}
+
+export interface DefaultValue {
+  readonly kind: 'DB_GENERATED' | 'TRANSFORMER_GENERATED';
+  readonly value: string | number;
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:25]: boolean/object?

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  public setFields(fields: string[]): void {
+    this.fields = [...fields];
+  }
+}
+
+export class Model {
+  private fields: Field[] = [];
+  private primaryKey: Index | undefined = undefined;
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:50]: is undefined explicit assignment needed here?

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  constructor(private name: string) {
+  }
+
+  public getName(): string {
+    return this.name;
+  }
+
+  public addField(field: Field): void {
+    if (this.hasField(field.name)) {
+      throw new Error("Field already exists.");
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:62]: field name in error?

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  public setPrimaryKey(fields: string[]): void {
+    const primaryKey = new Index('PRIMARY_KEY');
+    primaryKey.setFields(fields);
+    this.primaryKey = primaryKey;
+  }
+
+  public addIndex(name: string, fields: string[]): void {
+    if (this.hasIndex(name)) {
+      throw new Error("Index already exists.");
```
Comment (alharris-at) [inline packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:79]: index name in the error :D

**Review by alharris-at — COMMENTED** [review]
Mostly minor comments/questions.

File: packages/amplify-graphql-schema-generator/src/schema-representation/engine.ts (hunk 1 above)
Comment (sundersc) [reply packages/amplify-graphql-schema-generator/src/schema-representation/engine.ts:1]: This value is intended to determine the queries required to retrieve table metadata. The logic differs even between relational databases, so it makes sense to have the actual DB engine names.

File: packages/amplify-graphql-schema-generator/src/schema-representation/types.ts (hunk 5 above)
Comment (sundersc) [reply packages/amplify-graphql-schema-generator/src/schema-representation/types.ts:50]: Yes, it is helpful when we use strict typing.

**Review by alharris-at — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Pull request merged**